contracts/SplitMerge.sol

```diff
@@ -101,135 +121,148 @@ contract SplitMerge is Initializable, Ownable {
+    tokenIdToSplitOperations[_spaceTokenId].push(address(newSplitOperation));
+    allSplitOperations.push(newSplitOperation);
+
+    spaceToken.transferFrom(spaceTokenOwner, address(newSplitOperation), _spaceTokenId);
+    newSplitOperation.init();
+
+    emit SplitOperationStart(_spaceTokenId, address(newSplitOperation));
+    return newSplitOperation;
+  }
+
+  function finishSplitOperation(uint256 _spaceTokenId) public {
+    require(tokenIdToSplitOperations[_spaceTokenId].length > 0, "Split operations for this token not exists");
+    address splitOperationAddress = tokenIdToSplitOperations[_spaceTokenId][tokenIdToSplitOperations[_spaceTokenId].length - 1];
+    require(activeSplitOperations[splitOperationAddress], "Method should be called from active SpaceSplitOperation contract");
+    SpaceSplitOperation splitOperation = SpaceSplitOperation(splitOperationAddress);
+
+    (uint256[] memory subjectContourOutput, address subjectTokenOwner, uint256 resultContoursLength) = splitOperation.getFinishInfo();
+
+    packageToContour[_spaceTokenId] = subjectContourOutput;
+
+    int256 minHeight = packageToHeights[_spaceTokenId][0];
+
+    int256[] memory subjectPackageHeights = new int256[](packageToContour[_spaceTokenId].length);
+    for (uint i = 0; i < packageToContour[_spaceTokenId].length; i++) {
+      if (i + 1 > packageToHeights[_spaceTokenId].length) {
+        subjectPackageHeights[i] = minHeight;
+      } else {
+        if (packageToHeights[_spaceTokenId][i] < minHeight) {
+          minHeight = packageToHeights[_spaceTokenId][i];
         }
+        subjectPackageHeights[i] = packageToHeights[_spaceTokenId][i];
       }
     }
 
-    for (uint i = 0; i < checkContour1.length + checkContour2.length; i++) {
-      uint256 el = 0;
-      if (i < checkContour1.length) {
-        if (checkContour1[i] != 0) {
-          el = checkContour1[i];
-        }
-      } else if (checkContour2[i - checkContour1.length] != 0) {
-        el = checkContour2[i - checkContour1.length];
-      }
+    packageToHeights[_spaceTokenId] = subjectPackageHeights;
+    spaceToken.transferFrom(splitOperationAddress, subjectTokenOwner, _spaceTokenId);
+    for (uint j = 0; j < resultContoursLength; j++) {
+      uint256 newPackageId = spaceToken.mint(subjectTokenOwner);
+      packageToContour[newPackageId] = splitOperation.getResultContour(j);
 
-      if (el != 0) {
-        int index = ArrayUtils.uintFind(sourceContour, el);
-        require(index != - 1, "Unique element not exists in source contour");
-        sourceContour[uint(index)] = 0;
+      for (uint k = 0; k < packageToContour[newPackageId].length; k++) {
+        packageToHeights[newPackageId].push(minHeight);
       }
+      packageToLevel[newPackageId] = getPackageLevel(_spaceTokenId);
+      emit NewSpaceToken(newPackageId);
     }
+
+    activeSplitOperations[splitOperationAddress] = false;
+  }
+
+  function cancelSplitPackage(uint256 _spaceTokenId) public {
```

Comment:
Do we care about permission checks at the moment?

Reply:
yes, we are, fixed